Add IPv6 address of metadata service to sg rules

sven-rosenzweig opened the pull request:
So far only DHCPv6 and ICMPv6 traffic is refelcted in the default infrastructure rules of NSX-T. The IPv6 address of the metadata service requires clearing as well.

In order to reflect those changes in NSX-T, it requires a check if the rule configuration has changed. So far only creation of missing policies is supported.

sven-rosenzweig committed:
Support updating default security policies

Creating the default security policies (DHCP, ICMP and Metadata
allow), happens on driver start (in case they have not been realized so far).
Changing those was not supported.

Before starting the driver, it is now checked if the realized rules
differs from the rules defined in NSX-T constants. If a change is
detected, the rules defined in the NSX-T constants will be realized as
defined.

sebageek commented on networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/constants_nsx.py:
```diff
@@ -233,7 +233,7 @@
                 "display_name": "HTTP",
                 "source_groups": ["ANY"],
-                "destination_groups": ["169.254.169.254"],
+                "destination_groups": ["169.254.169.254", "fe80::a9fe:a9fe"],
```
If you like constants we have `neutron_lib.constants.METADATA_V6_IP`. There's also one for v4.

sebageek commented on networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py:
```diff
@@ -635,13 +635,27 @@ def _setup_default_infrastructure_rules(self):
             res = self.client.get(path=path)
             if res.ok:
-                continue
+                if self._check_for_infrastructure_changes():
```
What is this method? I haven't found it in the source. Do you want to call `_check_infrastructure_rules_for_updates()` here?

sebageek commented on networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py:
```diff
@@ -635,13 +635,27 @@ def _setup_default_infrastructure_rules(self):
 
+    def _check_infrastructure_rules_for_updates(self, policies_from_cfg, security_policies):
+        for rule_from_cfg in policies_from_cfg['rules']:
+            for rule_policy in sg_policy["rules"]:
```
What is `sg_policy`? is that `security_policies` but with a different name?

sebageek commented on networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py:
```diff
@@ -635,13 +635,27 @@ def _setup_default_infrastructure_rules(self):
+        for rule_from_cfg in policies_from_cfg['rules']:
+            for rule_policy in sg_policy["rules"]:
+                if rule_policy['id'] == rule_from_cfg['id']:
+                    shared_keys = set(rule.keys()).intersection(set(rule_policy.keys()))
```
[optional] I like using operators for set operations, but it depends of what you like yourself. For intersection you have the option of using `&`, so `set(rule) & set(rule_policy)` (the `.keys()` is implicit for dict-to-list-conversion but that's also a choice on how verbose you like to be, makes it more ovious that you're dealing with a dict).

sebageek commented on networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py:
```diff
@@ -635,13 +635,27 @@ def _setup_default_infrastructure_rules(self):
+                if rule_policy['id'] == rule_from_cfg['id']:
+                    shared_keys = set(rule.keys()).intersection(set(rule_policy.keys()))
+                    diff = set(o for o in shared_keys if rule[o] != rule_policy[o])
+                    LOG.info("Infrastructure rule %s has changed in %s - new rule %s" % (rule_from_cfg, diff, rule_from_cfg))
```
I'd pass the format attributes as arguments to `LOG.info()` instead of doing the formatting beforehand.

sebageek commented on networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py:
```diff
@@ -635,13 +635,27 @@ def _setup_default_infrastructure_rules(self):
+                    LOG.info("Infrastructure rule %s has changed in %s - new rule %s" % (rule_from_cfg, diff, rule_from_cfg))
+                    return True
+                else:
+                    LOG.info("New infrastructure rule found %s" % rule_from_cfg)
```
Same here for the Log args.

joroaf commented on networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py:
```diff
@@ -635,13 +635,27 @@ def _setup_default_infrastructure_rules(self):
             if res.ok:
-                continue
+                if self._check_for_infrastructure_changes():
+                    self.client.put(path=path, data=policy).raise_for_status()
```
Should we update the default infra policy every time with the desired rules? What if the operator added his own rules? Is this the desired behaviour?

sven-rosenzweig committed:
Add IPv6 address of metadata service to sg rules

So far only DHCPv6 and ICMPv6 traffic is refelcted in the default
infrastructure rules of NSX-T. The IPv6 address of the metadata
service requires clearing as well.

sebageek commented on networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py:
```diff
@@ -113,6 +114,9 @@ class API(object):
 
     INFRA = "/policy/api/v1/infra"
 
+class Ignore(set):
```
[optional] Could get another name that denotes more what it does. `IgnoreGivenKeys`? `DiffIgnoreKeys`? You also could move it closer to where it's used scope wise, on the other hand maybe you'll need it some day and then you want it in that package directly accessible? Dunno. Both options make sense to me.

sebageek commented on networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_policy.py:
```diff
@@ -113,6 +114,9 @@ class API(object):
 
+class Ignore(set):
+    def __contains__(self, key):
+        return set.__contains__(self, key[1])
```
would `super().__contains__(key[1])` work as well?
